src/Nethermind/Nethermind.Evm/TransactionProcessing/SystemTxProcessor.cs

```diff
@@ -0,0 +1,512 @@
+
+        public SystemTxProcessor(
+            ISpecProvider? specProvider,
+            IWorldState? worldState,
+            IVirtualMachine? virtualMachine,
+            ILogManager? logManager)
+        {
+            _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+            _specProvider = specProvider ?? throw new ArgumentNullException(nameof(specProvider));
+            _worldState = worldState ?? throw new ArgumentNullException(nameof(worldState));
+            _virtualMachine = virtualMachine ?? throw new ArgumentNullException(nameof(virtualMachine));
+            _ecdsa = new EthereumEcdsa(specProvider.ChainId, logManager);
+        }
+
+        public void CallAndRestore(Transaction transaction, BlockHeader block, ITxTracer txTracer)
+        {
+            Execute(transaction, block, txTracer, ExecutionOptions.CommitAndRestore);
+        }
+
+        public void BuildUp(Transaction transaction, BlockHeader block, ITxTracer txTracer)
+        {
+            // we need to treat the result of previous transaction as the original value of next transaction
+            // when we do not commit
+            _worldState.TakeSnapshot(true);
+            Execute(transaction, block, txTracer, ExecutionOptions.None);
+        }
+
+        public void Execute(Transaction transaction, BlockHeader block, ITxTracer txTracer)
+        {
+            Execute(transaction, block, txTracer, ExecutionOptions.Commit);
+        }
+
+        public void Trace(Transaction transaction, BlockHeader block, ITxTracer txTracer)
+        {
+            Execute(transaction, block, txTracer, ExecutionOptions.NoValidation);
+        }
+
+        protected virtual void Execute(Transaction tx, BlockHeader header, ITxTracer tracer, ExecutionOptions opts)
+        {
+            IReleaseSpec spec = _specProvider.GetSpec(header);
+            if (!spec.IsEip4788Enabled) // ToDo add comment
+                spec = new SystemTransactionReleaseSpec(spec);
+
+            // restore is CallAndRestore - previous call, we will restore state after the execution
+            bool restore = opts.HasFlag(ExecutionOptions.Restore);
+            // commit - is for standard execute, we will commit thee state after execution
+            // !commit - is for build up during block production, we won't commit state after each transaction to support rollbacks
+            // we commit only after all block is constructed
+            bool commit = opts.HasFlag(ExecutionOptions.Commit) || !spec.IsEip658Enabled;
+
+            if (!ValidateStatic(tx, header, spec, tracer, opts, out long intrinsicGas))
+                return;
+
+            UInt256 effectiveGasPrice =
+                tx.CalculateEffectiveGasPrice(spec.IsEip1559Enabled, header.BaseFeePerGas);
+
+            if (opts == ExecutionOptions.Commit || opts == ExecutionOptions.None)
+            {
+                decimal gasPrice = (decimal)effectiveGasPrice / 1_000_000_000m;
```

Comment:
I guess you could reuse more code. Let's discuss it.

Reply:
Yes, I am planning on reusing as much as possible from code, just trying to find a way that doesn't make the execution life cycle abit convoluted